frontend/dockerfile/dockerfile_heredoc_test.go

```diff
@@ -0,0 +1,488 @@
+		LocalDirs: map[string]string{
+			builder.DefaultLocalNameDockerfile: dir,
+			builder.DefaultLocalNameContext:    dir,
+		},
+	}, nil)
+	require.NoError(t, err)
+
+	dt, err := ioutil.ReadFile(filepath.Join(destDir, "dest"))
+	require.NoError(t, err)
+	require.Equal(t, "single file\n", string(dt))
+
+	dt, err = ioutil.ReadFile(filepath.Join(destDir, "destdir/EOF"))
+	require.NoError(t, err)
+	require.Equal(t, "first file\n", string(dt))
+
+	dt, err = ioutil.ReadFile(filepath.Join(destDir, "destdir/EOF2"))
+	require.NoError(t, err)
+	require.Equal(t, "second file\n", string(dt))
+}
+
+func testRunBasicHeredoc(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+
+	dockerfile := []byte(`
+FROM busybox AS build
+
+RUN <<EOF
+echo "i am" >> /dest
+whoami >> /dest
+EOF
+
+FROM scratch
+COPY --from=build /dest /dest
+`)
+
+	dir, err := tmpdir(
+		fstest.CreateFile("Dockerfile", []byte(dockerfile), 0600),
+	)
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	c, err := client.New(context.TODO(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	destDir, err := ioutil.TempDir("", "buildkit")
+	require.NoError(t, err)
+	defer os.RemoveAll(destDir)
+
+	_, err = f.Solve(context.TODO(), c, client.SolveOpt{
+		Exports: []client.ExportEntry{
+			{
+				Type:      client.ExporterLocal,
+				OutputDir: destDir,
+			},
+		},
+		LocalDirs: map[string]string{
+			builder.DefaultLocalNameDockerfile: dir,
+			builder.DefaultLocalNameContext:    dir,
+		},
+	}, nil)
+	require.NoError(t, err)
+
+	dt, err := ioutil.ReadFile(filepath.Join(destDir, "dest"))
+	require.NoError(t, err)
+	require.Equal(t, "i am\nroot\n", string(dt))
+}
+
+func testRunFakeHeredoc(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+
+	dockerfile := []byte(`
+FROM alpine AS build
+
+RUN apk add python3 --no-cache
+
+SHELL ["/usr/bin/python3", "-c"]
+RUN <<EOF
+with open("/dest", "w") as f:
+    print("foo", file=f)
+EOF
+
+FROM scratch
+COPY --from=build /dest /dest
+`)
+
+	dir, err := tmpdir(
+		fstest.CreateFile("Dockerfile", []byte(dockerfile), 0600),
+	)
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	c, err := client.New(context.TODO(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	destDir, err := ioutil.TempDir("", "buildkit")
+	require.NoError(t, err)
+	defer os.RemoveAll(destDir)
+
+	_, err = f.Solve(context.TODO(), c, client.SolveOpt{
+		Exports: []client.ExportEntry{
+			{
+				Type:      client.ExporterLocal,
+				OutputDir: destDir,
+			},
+		},
+		LocalDirs: map[string]string{
+			builder.DefaultLocalNameDockerfile: dir,
+			builder.DefaultLocalNameContext:    dir,
+		},
+	}, nil)
+	require.NoError(t, err)
+
+	dt, err := ioutil.ReadFile(filepath.Join(destDir, "dest"))
+	require.NoError(t, err)
+	require.Equal(t, "foo\n", string(dt))
+}
+
+func testRunShebangHeredoc(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+
+	dockerfile := []byte(`
+FROM alpine AS build
+
+RUN apk add python3 --no-cache
```

Comment:
Aha! I think I've done better than making a "fake-shell" - I've just used awk. Since it's part of busybox, and it's definitely a non-traditional shell.